### PR TITLE
Build it

### PR DESCRIPTION
**Pull request by ctrlShiftBryan**
This is a quick prototype wrapping vue-select into a component that consuming an api endpoint using ajax for searching ICD10 codes. 

A test dev server is available in the server folder.

### CODE REVIEW

File: src/App.vue [hunk 1]
```diff
@@ -1,17 +1,19 @@
 <template>
   <div id="app">
-    <img alt="Vue logo" src="./assets/logo.png">
-    <HelloWorld msg="Welcome to Your Vue.js App"/>
+    <IcdSelect dataUrl="http://localhost:8082/icd10" v-model="value"/>
```
Comment (dannyrb) [inline src/App.vue:3]: I get that you're just stubbing this out as a POC, but I'll leave some unwanted advice sprinkled in anyway:

- `dataUrl` (or at least the base URL) could be an environment variable:
    - https://cli.vuejs.org/guide/mode-and-env.html#modes

File: src/components/IcdSelect.vue [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    return {
+      localSelected: null,
+      options: []
+    };
+  },
+  props: {
+    dataUrl: {},
+    value: {}
+  },
+  components: {
```
Comment (dannyrb) [inline src/components/IcdSelect.vue]: Your ordering for top level Vue Component properties is a little odd. You can see the recommended ordering here:

https://vuejs.org/v2/style-guide/#Component-instance-options-order-recommended

We tend to use [`eslint-plugin-vue`](https://github.com/vuejs/eslint-plugin-vue) and autofix on save.

VS Code gives you the ability to set recommended plugins for a solution, as well as workspace settings that enable/disable features for the project, regardless of who on the team pulled it down.

File: src/components/IcdSelect.vue [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    "v-select": vSelect
```
Comment (dannyrb) [inline src/components/IcdSelect.vue]: While this works, `vSelect` should already map to `v-select`. You shouldn't need to provide a new name as a string unless you want the imported component's name to differ from how you would like to reference the component in the markup.

File: src/components/IcdSelect.vue [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+      if (search.length >= 3) {
+        loading(true);
+        this.search(loading, search, this);
+      }
+    },
+    search: (loading, search, vm) => {
+      fetch(`${vm.dataUrl}?q=${escape(search)}`).then(res => {
+        res.json().then(json => (vm.options = json.data));
+        loading(false);
+      });
```
Comment (dannyrb) [inline src/components/IcdSelect.vue]:
```
async search(loading, queryTerm) {
    try {
        const response = await fetch(`${this.dataUrl}?q=${escape(queryTerm)}`);
        const json = await response.json();
        this.options = json.data
    } finally {
        loading(false);
    }
}
```

`async`/`await` makes me happy 🤷‍♂️ 

You shouldn't need to pass the `vm` here?
Also, if you were using `axios`, leveraging interceptors to handle different response codes is 👌 🔥

File: src/components/IcdSelect.vue [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    }
+  },
+  created: function() {
+    // the initial value is just the value with no label
+    // so we must make and ajax call to get the label too
+    fetch(`${this.dataUrl}?q=${escape(this.value)}`, {
+      // credentials: "include"
+    }).then(res => {
+      res.json().then(json => (this.localSelected = json.data[0]));
+    });
```
Comment (dannyrb) [inline src/components/IcdSelect.vue]: I think you'd want to 💣 💥 this post-POC. Ideally, this is something that's already fluffed and ready from the server (or somewhere else) when the component gets this data.

File: src/components/IcdSelect.vue [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  components: {
+    "v-select": vSelect
+  },
+  methods: {
+    updateSelected(newValue) {
+      this.$emit("input", newValue.code);
+    },
+    onSearch(search, loading) {
+      if (search.length >= 3) {
+        loading(true);
```
Comment (dannyrb) [inline src/components/IcdSelect.vue:51]: Might as well move this to the `search` method, so the toggling logic is together.

File: src/components/IcdSelect.vue [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+  name: "HelloWorld",
+  data: function() {
+    return {
+      localSelected: null,
+      options: []
+    };
+  },
+  props: {
+    dataUrl: {},
+    value: {}
```
Comment (dannyrb) [inline src/components/IcdSelect.vue:39]: You'll, at a minimum, want to expand these a bit to limit what they can be and do.

```js
value: {
    type: String,
    required: true
}
```

**Review by dannyrb — COMMENTED** [review]
> What I’m really wondering is if you have a component that supports v-model and you want to write a component with that inside so that the wrapping component also supports v-model, can you just do something like `<Parent v-model="value" />` and then in the child `<Child v-model="" />`?

I think this would be doable if you went with the deconstructed version of `v-model`:
```js
:value="someProp"
@input="updateSelected"
```

And triggered another `input` event in the handler, and continued to pass the value up.

> Also in this case the ‘local’ value in the child is actually a object not a string so the design I came up with is probably best.

That does make all of this a bit hairy. It's hard to `sync` values with `v-model` when the value you need "locally" and what your VM/Parent cares about are different. Maybe keep the selected value in the same format to make this easier to play with, but then use a fancy getter or a method to extract only the values you need and in the format you expect in your VM.

Another approach is using the `value` prop as an `initial` value for `localSelected`:

```
localSelected: this.value,
```

If you want the parent to be able to trigger an update for this component, `watch` the prop, and update your `localSelected` when it changes to a value that is not already selected.


When/if you add in `vuex`, and realize that `v-model` on a `state` value is a pain, because you're not allowed to mutate state, `computed setters` can be helpful:

```js
computed: {
  someValue: {
    get () { return this.$store.state.someValue},
    set (v) { this.$store.commit('SET_SOME_VALUE', v) }
  }
}
```

**tl;dr**

Your approach is fine, especially given the different representations. It's probably worth making your component dumber, and letting a different component/layer handling the mapping between the "selected value" binding format, and the "value" you actually care about. As always, lots of interesting ways to approach this.

File: src/components/IcdSelect.vue (hunk 2 above)
Comment (ctrlShiftBryan) [reply src/components/IcdSelect.vue]: 👍 yeah in the real project we just have eslint setup with plugin:vue/recommended where this cli version doesn't seem to have it, so I don;t even think about it.

File: src/components/IcdSelect.vue (hunk 3 above)
Comment (ctrlShiftBryan) [reply src/components/IcdSelect.vue]: 👍 thanks this felt weird

File: src/components/IcdSelect.vue (hunk 6 above)
Comment (ctrlShiftBryan) [reply src/components/IcdSelect.vue:51]: Yeah hmmm didn't even think about this after I copy pasted from the official docs.

File: src/components/IcdSelect.vue (hunk 5 above)
Comment (ctrlShiftBryan) [reply src/components/IcdSelect.vue]: Yeah that was my initial thought too. Porting this from back/select2 and thats how they did but now that you mention it I don't like the extra ajax calls on pageload.

File: src/App.vue (hunk 1 above)
Comment (ctrlShiftBryan) [reply src/App.vue:3]: Kinda wishing I used cli in the main project. I'm a little torn though. When I first played around with it was at version 2 and different combinations of options for testing and linting and it didn't actually work. Like I couldn't get the default tests to pass and stuff. Also not really a fan of having webpack abstracted away like it is.

This was the first project I played with v3 and it has all just worked out of the box unlike v2.

File: src/components/IcdSelect.vue (hunk 7 above)
Comment (ctrlShiftBryan) [reply src/components/IcdSelect.vue:39]: 👍 yeah main project that won't get past our linter. eslint is so impressive!

File: src/App.vue (hunk 1 above)
Comment (dannyrb) [reply src/App.vue:3]: > not really a fan of having webpack abstracted away like it is

I get why they did it, but it does feel overly cumbersome. WebPack can already be difficult to wrestle with, burying it a little deeper makes troubleshooting where exactly wires are getting crossed that much more difficult.

FWIW, they do have some commands that output the "vanilla" webpack config that's used for each build type.
